src/erlcloud_aws.erl

```diff
@@ -264,23 +264,106 @@ format_timestamp({{Yr, Mo, Da}, {H, M, S}}) ->
-        _ -> default_config_profile()
+    case config_env() of
+        {ok, Config} -> Config;
+        {error, _} -> #aws_config{}
+    end.
+
+
+%%%---------------------------------------------------------------------------
+-spec auto_config() -> {ok, aws_config()} | undefined.
+%%%---------------------------------------------------------------------------
+%% @doc Generate config using the best available credentials
+%%
+%% This function will generate a valid <code>#aws_config{}</code> based on
+%% the best available credentials source in the current environment.  The
+%% following sources of credentials will be used, in order:
+%%
+%% <ol>
+%%   <li>Environment Variables
+%%     <p>An Id, Key and optionally Token, will be sourced from the
+%%     environment variables <code>AWS_ACCESS_KEY_ID</code>,
+%%     <code>AWS_SECRET_ACCESS_KEY</code> and
+%%     <code>AWS_SECURITY_TOKEN</code> respectively.  Both the Id and Key
+%%     values must e non-empty for this form of credentials to be considered
+%%     valid.</p>
+%%   </li>
+%%
+%%   <li>User Profile
+%%     <p>The default user profile credentials will be sourced using the
+%%     {@link profile/0} function, if available for the current user.</p>
+%%   </li>
+%%
+%%   <li>Host Metadata
+%%     <p>The credentials available via host metadata will be sourced, if
+%%     available.</p>
+%%   </li>
+%% </ol>
+%%
+%% If none if these credential sources are available, this function will
+%% return <code>undefined</code>.
+%%
+auto_config() ->
+    case config_env() of
+        {ok, _Config} = Result -> Result;
+        {error, _} -> auto_config_profile()
     end.
 
-default_config_profile() ->
+auto_config_profile() ->
     case profile() of
-        {ok, Config} -> Config;
-        _ -> #aws_config{}
+        {ok, _Config} = Result -> Result;
+        {error, _} -> auto_config_metadata()
+    end.
+
+auto_config_metadata() ->
+    case config_metadata() of
+        {ok, _Config} = Result -> Result;
+        {error, _} -> undefined
+    end.
+
+
+config_env() ->
+    case {os:getenv("AWS_ACCESS_KEY_ID"), os:getenv("AWS_SECRET_ACCESS_KEY"),
+          os:getenv("AWS_SECURITY_TOKEN")} of
+        {KeyId, Secret, T} when is_list(KeyId), is_list(Secret) ->
+            Token = if is_list(T) -> T; true -> undefined end,
+            Config = #aws_config{access_key_id = KeyId,
+                                 secret_access_key = Secret,
+                                 security_token = Token},
+            {ok, Config};
+        _ -> {error, environment_config_unavailable}
+    end.
+
+config_metadata() ->
+    Config = #aws_config{},
+    case get_metadata_credentials( Config ) of
+        {ok, #metadata_credentials{
+                access_key_id = Id,
+                secret_access_key = Secret,
+                security_token = Token,
+                expiration_gregorian_seconds = GregorianSecs}} ->
+            EpochTimeout = GregorianSecs - 62167219200,
```

Comment:
Could you please define this Unix epoch to gregorian seconds constant?
I think it may be useful somewhere else.